### DISCUSSION TITLE
Firebase Storage: Unit testing rules with test report

### DISCUSSION THREAD

**Discussion by IchordeDionysos** [firebase/firebase-js-sdk#4969] [General]
I think it's very nice that we now have a Firebase Storage emulator 😍 

What I would like to have added there would be the possibility to test the rules similar to how the Firestore emulator rules can be tested!

@samtstern already started work on it, but it's not yet fully ready and still fails with a few errors.
https://github.com/firebase/firebase-tools/pull/3332
https://github.com/firebase/firebase-tools/pull/3337
https://github.com/firebase/firebase-js-sdk/pull/4863

There is a related issue with using `@firebase/rules-unit-testing` with the storage emulator:
https://github.com/firebase/firebase-tools/issues/3442

What I would like to concretely see is:

- [Storage support in @firebase/rules-unit-testing][rules testing]
- [Generating test reports similar to Firestore][test reports]

I would like to add support for generating Firebase Storage coverage reports to a [package (`@simpleclub/firebase-rules-coverage`)][rules coverage] a package I've developed that transforms the test report into an LCOV coverage file that can be uploaded to a generic test coverage tool.

[test reports]: https://firebase.google.com/docs/firestore/security/test-rules-emulator#generate_test_reports
[rules testing]: https://www.npmjs.com/package/@firebase/rules-unit-testing
[rules coverage]: https://www.npmjs.com/package/@simpleclub/firebase-rules-coverage

**Comment by samtstern** [comment]
@IchordeDionysos thanks for starting this discussion! Just curious why this is filed on `firebase-admin-node` since Admin SDK doesn't interact with Rules. This would seem better suited for `firebase-js-sdk` or `firebase-tools`.

Happy to transfer it if you agree.

**Reply by IchordeDionysos** [reply]
Uhh yes makes sense 😄
Also my other discussion probably!
